Verify `private/topology`

henriman opened the pull request:
Still contains workarounds for bugs:
- ([831](https://github.com/viperproject/gobra/issues/831)) Byte slice of low string is low
- ([832](https://github.com/viperproject/gobra/issues/832)) String of low byte slice is low; see `AssumeLowSliceToLowString` in `verification/utils/sif/assumptions.gobra`
- ([835](https://github.com/viperproject/gobra/issues/835)/[890](https://github.com/viperproject/gobra/issues/890)) Slice created from low values is low

TODO
- Align contract clauses (I haven't done so yet in order to make the diff more useful)
- Use `// @` instead of `//@`?

jcp19 commented on private/topology/linktype.go:
```diff
@@ -59,51 +61,78 @@ func (l LinkType) String() string {
 func LinkTypeFromString(s string) (res LinkType) {
 	var l /*@@@*/ LinkType
 	tmp := []byte(s)
+	// TODO: Once Gobra issue #831 is resolved, remove this assumption.
+	//@ assume forall i int :: { &tmp[i] } 0 <= i && i < len(tmp) ==> low(tmp[i])
```
Is this assumption still necessary? (same for all assumptions like this below)

henriman replied:
Gobra issues #831 and #832 are not resolved by adding `&& low(i) ==> ...`, therefore these assumptions are still necessary.

However, it turns out that I could remove some postconditions from `LinkType.MarshalText` which made the assumptions in that method unnecessary.

I will also add `&& low(i) ==>` where appropriate.

jcp19 commented on private/topology/linktype.go:
```diff
@@ -59,51 +61,78 @@ func (l LinkType) String() string {
 // @ ensures err == nil ==> sl.Bytes(res, 0, len(res))
+// @ ensures err == nil ==> low(len(res)) && 
+// @ 	forall i int :: { sl.GetByte(res, 0, len(res), i) } 0 <= i && i < len(res) ==>
+// @ 	low(sl.GetByte(res, 0, len(res), i))
```
we can now have some kind of opaque hyper function for this (here and in some other places of this PR)

henriman replied:
As discussed, I will bundle the addition of such functions (for all packages up to and including `pkg/slayers/path`) in PR #10.

jcp19 commented on private/topology/linktype.go:
```diff
@@ -59,51 +61,78 @@ func (l LinkType) String() string {
 // @ ensures err == nil ==> sl.Bytes(res, 0, len(res))
+// @ ensures err == nil ==> low(len(res)) && 
+// @ 	forall i int :: { sl.GetByte(res, 0, len(res), i) } 0 <= i && i < len(res) ==>
+// @ 	low(sl.GetByte(res, 0, len(res), i))
```
```suggestion
// @ 		low(sl.GetByte(res, 0, len(res), i))
```

henriman commented:
@jcp19: besides addressing your feedback, I have added `&& low(i) ==>` in quantified assertions where appropriate, and adapted the beginning of `(*LinkType).UnmarshalText` to not require using `BeforeUnfold:`/`old[BeforeUnfold](...` (to make it cleaner, IMO)

jcp19 commented on private/topology/linktype.go:
```diff
@@ -59,17 +61,21 @@ func (l LinkType) String() string {
 func LinkTypeFromString(s string) (res LinkType) {
 	var l /*@@@*/ LinkType
 	tmp := []byte(s)
+	// TODO: Once Gobra issue #831 is resolved, remove this assumption.
+	//@ assume forall i int :: { &tmp[i] } 0 <= i && i < len(tmp) && low(i) ==> low(tmp[i])
```
to avoid the proliferation of these kinds of assumptions throughout the codebase (regardless of how benign they are), I would rather introduce a lemma here that gives me this postcondition.

henriman replied:
Using your fix (PR [#961](https://github.com/viperproject/gobra/pull/961)), I was able to remove this assumption

jcp19 replied:
cool! FYI, I just merged PR 961 in the PR that fixes issue 846

jcp19 commented:
LGTM. Please address my comment and then you can merge this

jcp19 commented:
Is there any issue blocking the merging of this or the other PRs?
